src/main/java/org/opentripplanner/routing/alertpatch/EntitySelector.java

```diff
@@ -164,16 +192,23 @@ public boolean equals(Object o) {
       if (!stop.equals(that.stop)) {
         return false;
       }
-      return routeOrTrip.equals(that.routeOrTrip);
+
+      if (!routeOrTrip.equals(that.routeOrTrip)) {
+        return false;
+      }
+
+      return serviceDate != null ? serviceDate.equals(that.serviceDate) : that.serviceDate == null;
     }
 
     @Override
     public int hashCode() {
       if (hash == 0) {
-        int result = stop.hashCode();
-        hash = 31 * result + routeOrTrip.hashCode();
+        int stopResult = stop.hashCode();
+        int serviceDateResult = serviceDate == null ? 0:serviceDate.hashCode();
+        hash = 31 * stopResult + 31 * serviceDateResult + routeOrTrip.hashCode();
```

Comment:
I would prefer:
```java
hash = Objects.hash(stop.hashCode(), serviceDate.hashCode(), routeOrTrip.hashCode());
```

Reply:
I set this in the constructor: `this.hash = Objects.hash(stop, serviceDate, routeOrTrip);` and made the `hash` final. This make it thread-safe.